### PR TITLE
Fix elided lifetimes shown as `'_` on async functions

### LINKED ISSUE #63037: rustdoc shows anonymous elided lifetimes in async fn signature
When run on an `async fn` with elided lifetimes, `rustdoc` generates documentation containing the anonymous elided lifetimes in the function's signature. Running `rustdoc` on the blocking variant of the same function produces a correct signature.

You can reproduce this issue by generating docs for the following code:
```rust
#![feature(async_await)]

pub async fn asynchronous(foo: &str, bar: &str) -> String {
    format!("{}{}", foo, bar)
}

pub fn blocking(foo: &str, bar: &str) -> String {
    format!("{}{}", foo, bar)
}
```

The asynchronous function's docs show the elided lifetimes, the blocking function's docs don't:

```rust
pub async fn asynchronous<'_, '_>(foo: &'_ str, bar: &'_ str) -> String
```

```rust
pub fn blocking(foo: &str, bar: &str) -> String
```

## Meta

`rustdoc --version --verbose`:
```
rustdoc 1.38.0-nightly (c43753f91 2019-07-26)
binary: rustdoc
commit-hash: c43753f910aae000f8bcb0a502407ea332afc74b
commit-date: 2019-07-26
host: x86_64-unknown-linux-gnu
release: 1.38.0-nightly
LLVM version: 9.0
```

### PR DESCRIPTION
**Pull request by jyn514**
Closes https://github.com/rust-lang/rust/issues/63037.

r? @tmandry on the implementation, @Darksonn on the test cases.

### CODE REVIEW

File: src/test/rustdoc/async-fn.rs [hunk 1]
```diff
@@ -52,3 +52,31 @@ pub trait Trait<const N: usize> {}
...
+pub async fn static_trait(foo: &str) -> Box<dyn Bar> {}
+// @has async_fn/fn.lifetime_for_trait.html
+// @has - '//pre[@class="rust fn"]' "pub async fn lifetime_for_trait(foo: &str) -> Box<dyn Bar + '_>"
+pub async fn lifetime_for_trait(foo: &str) -> Box<dyn Bar + '_> {}
+
+struct AsyncFdReadyGuard<'a, T> { x: &'a T }
+
+impl Foo {
+    // @has async_fn/struct.Foo.html
+    // @has - '//h4[@class="method"]' 'pub async fn complicated_lifetimes( &self, context: &impl Bar) -> impl Iterator<Item = &usize>'
```
Comment (jyn514) [inline src/test/rustdoc/async-fn.rs:77]: The test suite normalizes this really strangely; the original is 
image

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -1433,7 +1445,16 @@ impl Clean<Type> for hir::Ty<'_> {
             TyKind::Never => Never,
             TyKind::Ptr(ref m) => RawPointer(m.mutbl, box m.ty.clean(cx)),
             TyKind::Rptr(ref l, ref m) => {
-                let lifetime = if l.is_elided() { None } else { Some(l.clean(cx)) };
+                // There are two times a `Fresh` lifetime can be created:
+                // 1. For `&'_ x`, written by the user. This corresponds to `lower_lifetime` in `rustc_ast_lowering`.
+                // 2. For `&x` as a parameter to an `async fn`. This corresponds to `elided_ref_lifetime in `rustc_ast_lowering`.
+                //    See commit 749349fc9f7b12f212bca9ba2297e463328cb701 for more information.
```
Comment (tmandry) [inline src/librustdoc/clean/mod.rs:1451]: You can point straight to the PR #59286 which has the same description (it's shorter than a commit hash).

File: src/librustdoc/clean/mod.rs [hunk 3]
```diff
@@ -1433,7 +1445,16 @@ impl Clean<Type> for hir::Ty<'_> {
...
+                // Ideally we would only hide the `'_` for case 2., but I don't know a way to distinguish it.
+                // Turning `fn f(&'_ self)` into `fn f(&self)` isn't the worst thing in the world, though;
+                // there's no case where it could cause the function to fail to compile.
```
Comment (tmandry) [inline src/librustdoc/clean/mod.rs:1454]: Yeah, I think this is fine. What we don't want to eliminate is something like `Foo<'_>`, but you have a test that covers that already in the output position. Could you also add one for the input position?

File: src/test/rustdoc/async-fn.rs [hunk 4]
```diff
@@ -52,3 +53,31 @@ pub trait Trait<const N: usize> {}
 // @has async_fn/fn.const_generics.html
 // @has - '//pre[@class="rust fn"]' 'pub async fn const_generics<const N: usize>(_: impl Trait<N>)'
 pub async fn const_generics<const N: usize>(_: impl Trait<N>) {}
+
+// test that elided lifetimes are properly elided and not displayed as `'_`
```
Comment (tmandry) [inline src/test/rustdoc/async-fn.rs:57]: Can you add some tests with named lifetimes to show that we don't throw them out?

(It's pretty clear from the implementation that we won't, but this is future proofing!)

File: src/librustdoc/clean/mod.rs [hunk 5]
```diff
@@ -1433,7 +1445,16 @@ impl Clean<Type> for hir::Ty<'_> {
...
+                let elided =
+                    l.is_elided() || matches!(l.name, LifetimeName::Param(ParamName::Fresh(_)));
```
Comment (tmandry) [inline src/librustdoc/clean/mod.rs:1456]: I'm surprised that there are two possible ways we need to handle, when is it one versus the other? (Which test cases fail without one of these?)

Comment (jyn514) [reply src/librustdoc/clean/mod.rs:1456]: `Fresh` params aren't considered elided in rustc itself: https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_hir/hir.rs.html#130-143. Almost every test fails without this, the only thing `GenericParam` changes is removing the bogus `fn f<'_>`.

If you notice above, it was already checking `is_elided`, just not `Fresh`.

File: src/librustdoc/clean/mod.rs (hunk 3 above)
Comment (jyn514) [reply src/librustdoc/clean/mod.rs:1454]: Good idea, done.

File: src/test/rustdoc/async-fn.rs (hunk 4 above)
Comment (jyn514) [reply src/test/rustdoc/async-fn.rs:59]: Done, good idea.

**Review by Darksonn — COMMENTED** [review]
The test-cases look good, although I note that there are very few cases with `&self` or `&mut self`, which is where I typically see the issue.

File: src/librustdoc/clean/mod.rs (hunk 5 above)
Comment (tmandry) [reply src/librustdoc/clean/mod.rs:1456]: Ah I see. That makes me wonder if we should push this change up to `is_elided()`, but that might have other side effects.

**Pull request merged**